src/dlr_relayvm.cc

```diff
@@ -0,0 +1,258 @@
+#include "dlr_relayvm.h"
+
+
+#include <stdlib.h>
+#include <fstream>
+#include <iterator>
+#include <numeric>
+
+using namespace dlr;
+
+void RelayVMModel::InitModelPath(std::vector<std::string> paths) {
+  path_ = std::make_unique<ModelPath>();
+  std::vector<std::string> paths_vec;
+  for (auto path : paths) {
+    ListDir(path, paths_vec);
+  }
+
+  for(auto path: paths_vec) {
+    if (!EndsWith(path, LIBDLR) && EndsWith(path, ".so")) {
+      path_->model_lib = path;
+    } else if (EndsWith(path, ".ro")) {
+      path_->relay_executable = path;
+    } else if (EndsWith(path, ".meta")) {
+      path_->metadata = path;
+    }
+  }
+
+  if (path_->model_lib.empty() || path_->relay_executable.empty() || path_->metadata.empty()) {
+    LOG(FATAL) << "Invalid model artifact!";
+  }
+}
+
+void RelayVMModel::SetupVMModule() {
+    tvm::runtime::Module lib = tvm::runtime::Module::LoadFromFile(path_->model_lib);
+    std::ifstream relay_ob(path_->relay_executable, std::ios::binary);
+    std::string code_data((std::istreambuf_iterator<char>(relay_ob)), std::istreambuf_iterator<char>());
+
+    vm_executable_ = std::make_shared<tvm::runtime::Module>(tvm::runtime::vm::Executable::Load(code_data, lib));
+    vm_ = tvm::runtime::make_object<tvm::runtime::vm::VirtualMachine>();
+    vm_->LoadExecutable((tvm::runtime::vm::Executable*)vm_executable_->get());
+    vm_module_ = std::make_shared<tvm::runtime::Module>(tvm::runtime::Module(vm_));
+
+    tvm::runtime::PackedFunc init = vm_module_->GetFunction("init");
+    init(1, 0);
+}
+
+void RelayVMModel::LoadMetadata() {
+  LoadJsonFromFile(path_->metadata, metadata_);
+}
+
+void RelayVMModel::FetchInputNodesData() {
```

Comment:
Where do we get input shapes? For VM we might need to check whether the actual input shape is consistent with original relay input shapes. For example, if we have a dynamic input with shape (1, Any()), when we set input in VM we need to check so that (1, 4) is valid and (2, 3) is invalid.

Reply:
Yes we can do that. We can only  validate that number of channels and dimensions are consistent.

Reply:
Yeah for dimensions with ```Any``` we don't need to check them. Only need to check dimensions with actual number.

Reply:
And rank must match as well.

Reply:
I agree. Although I think this check should be performed in tvm I think.

Reply:
Currently TVM doesn't have something like metadata to record all the necessary info for this task on runtime side. We can do it in DLR now.

Reply:
Sure. I will follow up with in another PR. Thanks